File: src/CalcViewModel/Common/Automation/NarratorNotifier.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// Implementation of the NarratorNotifier class.

#include "pch.h"
#include "NarratorNotifier.h"

using namespace CalculatorApp::ViewModel::Common::Automation;
using namespace Platform;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Automation;
using namespace Windows::UI::Xaml::Automation::Peers;

DependencyProperty ^ NarratorNotifier::s_announcementProperty;

NarratorNotifier::NarratorNotifier()
{
}

void NarratorNotifier::Announce(NarratorAnnouncement ^ announcement)
{
    if (NarratorAnnouncement::IsValid(announcement))
    {
        if (m_announcementElement == nullptr)
        {
            m_announcementElement = ref new Windows::UI::Xaml::Controls::TextBlock();
        }

        auto peer = FrameworkElementAutomationPeer::FromElement(m_announcementElement);
        if (peer != nullptr)
        {
            peer->RaiseNotificationEvent(announcement->Kind, announcement->Processing, announcement->Announcement, announcement->ActivityId);
        }
    }
}

void NarratorNotifier::RegisterDependencyProperties()
{
    s_announcementProperty = DependencyProperty::Register(
        L"Announcement",              // The name of the dependency property.
        NarratorAnnouncement::typeid, // The type of the dependency property.
        NarratorNotifier::typeid,     // The owner of the dependency property.
        ref new PropertyMetadata(
            nullptr, // Default value of the dependency property.
            ref new PropertyChangedCallback(OnAnnouncementChanged)));
}

void NarratorNotifier::OnAnnouncementChanged(_In_ DependencyObject ^ dependencyObject, _In_ DependencyPropertyChangedEventArgs ^ e)
{
    auto instance = safe_cast<NarratorNotifier ^>(dependencyObject);
    if (instance != nullptr)
    {
        instance->Announce(safe_cast<NarratorAnnouncement ^>(e->NewValue));
    }
}

File: src/Calculator.ViewModels/Common/Automation/NarratorNotifier.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Windows.UI.Xaml;
using Windows.UI.Xaml.Automation.Peers;
using Windows.UI.Xaml.Controls;

namespace CalculatorApp.ViewModel.Common.Automation
{
    public sealed class NarratorNotifier : DependencyObject
    {
        private UIElement _announcementElement;

        private static DependencyProperty s_announcementProperty;

        public NarratorNotifier()
        {
        }

        public NarratorAnnouncement Announcement
        {
            get => GetAnnouncement(this);
            set => SetAnnouncement(this, value);
        }

        public static DependencyProperty AnnouncementProperty => s_announcementProperty;

        public static NarratorAnnouncement GetAnnouncement(DependencyObject element)
        {
            return (NarratorAnnouncement)element.GetValue(s_announcementProperty);
        }

        public static void SetAnnouncement(DependencyObject element, NarratorAnnouncement value)
        {
            element.SetValue(s_announcementProperty, value);
        }

        public static void RegisterDependencyProperties()
        {
            s_announcementProperty = DependencyProperty.Register(
                "Announcement",
                typeof(NarratorAnnouncement),
                typeof(NarratorNotifier),
                new PropertyMetadata(null, OnAnnouncementChanged));
        }

        public void Announce(NarratorAnnouncement announcement)
        {
            if (NarratorAnnouncement.IsValid(announcement))
            {
                if (_announcementElement == null)
                {
                    _announcementElement = new TextBlock();
                }

                var peer = FrameworkElementAutomationPeer.FromElement(_announcementElement);
                if (peer != null)
                {
                    peer.RaiseNotificationEvent(
                        announcement.Kind,
                        announcement.Processing,
                        announcement.Announcement,
                        announcement.ActivityId);
                }
            }
        }

        private static void OnAnnouncementChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs e)
        {
            if (dependencyObject is NarratorNotifier instance)
            {
                instance.Announce(e.NewValue as NarratorAnnouncement);
            }
        }
    }
}
